substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/RemoveRedundantClassInitClosure.java

```diff
@@ -0,0 +1,135 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.hosted.phases;
+
+import com.oracle.svm.hosted.meta.HostedMethod;
+import jdk.vm.ci.meta.JavaConstant;
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.collections.EconomicSet;
+import org.graalvm.collections.Equivalence;
+import org.graalvm.compiler.core.common.cfg.Loop;
+import org.graalvm.compiler.graph.Node;
+import org.graalvm.compiler.nodes.AbstractBeginNode;
+import org.graalvm.compiler.nodes.InvokeWithExceptionNode;
+import org.graalvm.compiler.nodes.ValueNode;
+import org.graalvm.compiler.nodes.cfg.Block;
+import org.graalvm.compiler.phases.graph.ReentrantBlockIterator;
+
+import java.util.List;
+
+/*
+* This class is used for processing one block of nodes, while ReentrantBlockIterator is iterating trough blocks of nodes.
+* Base purpose of this class is to ensure that for every class, which is initialized at runtime, DynamicHub.ensureInitialized
+* method is generated only once.
+* It's called from RemoveRedundantClassInitPhase.
+* */
+public class RemoveRedundantClassInitClosure extends ReentrantBlockIterator.BlockIteratorClosure<EconomicSet<JavaConstant>> {
+
+    private static final int WHICH_ARGUMENT_INDEX_IS_CLASS = 0;
+
+    private EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> nodesToBeRemoved;
+    private EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> nodesToRemain;
+
+    private final HostedMethod classInitMethod;
+
+    public RemoveRedundantClassInitClosure(HostedMethod classInitMethod) {
+        this.classInitMethod = classInitMethod;
+    }
+
+    public void removeRedundantNodes(RemoveRedundantClassInitClosure removeRedundantClassInitClosure, Block start) {
+        ReentrantBlockIterator.apply(removeRedundantClassInitClosure, start);
+    }
+
+    public EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> getNodesToBeRemoved() {
+        return nodesToBeRemoved;
+    }
+
+    public EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> getNodesToRemain() {
+        return nodesToRemain;
+    }
+
+    @Override
+    protected EconomicSet<JavaConstant> getInitialState() {
+        this.nodesToBeRemoved = EconomicMap.create(Equivalence.IDENTITY);
+        this.nodesToRemain = EconomicMap.create(Equivalence.IDENTITY);
+
+        return EconomicSet.create(Equivalence.IDENTITY);
```

Comment:
Why do you explicitly request `Equivalence.IDENTITY` for this set (and all the other sets/maps)? While all constants should be canonicalized, there is no need to restrict the optimization in case the canonicalization did not happen.

Reply:
Changed to default equivalence type.